crates/core/src/protocols/sumcheck/v3/bivariate_mlecheck.rs

```diff
@@ -0,0 +1,578 @@
+// Copyright 2025 Irreducible Inc.
+
+use std::{iter, mem, slice};
+
+use binius_compute::{
+	ComputeLayer, ComputeMemory, FSlice, FSliceMut, KernelBuffer, KernelMemMap, SizedSlice,
+	SlicesBatch,
+	alloc::{BumpAllocator, ComputeAllocator, HostBumpAllocator},
+};
+use binius_field::{
+	Field, TowerField,
+	util::{eq, powers},
+};
+use binius_math::{ArithExpr, CompositionPoly, EvaluationOrder, evaluate_univariate};
+use binius_utils::bail;
+
+use super::bivariate_product::{PhaseState, SumcheckMultilinear};
+use crate::{
+	composition::{BivariateProduct, IndexComposition},
+	protocols::sumcheck::{
+		CompositeSumClaim, EqIndSumcheckClaim, Error, RoundCoeffs, prove::SumcheckProver,
+	},
+};
+
+pub struct BivariateMLEcheckProver<'a, 'alloc, F: Field, Hal: ComputeLayer<F>> {
+	hal: &'a Hal,
+	dev_alloc: &'a BumpAllocator<'alloc, F, Hal::DevMem>,
+	host_alloc: &'a HostBumpAllocator<'a, F>,
+	n_vars_initial: usize,
+	n_vars_remaining: usize,
+	multilins: Vec<SumcheckMultilinear<'a, F, Hal::DevMem>>,
+	compositions: Vec<IndexComposition<BivariateProduct, 2>>,
+	last_coeffs_or_sums: PhaseState<F>,
+	eq_ind_prefix_eval: F,
+	// Wrapping it in Option is a temporary workaround for the lifetime problem
+	eq_ind_partial_evals: Option<FSliceMut<'a, F, Hal>>,
+	eq_ind_challenges: Vec<F>,
+	first_round_eval_1s: Option<Vec<F>>,
+}
+
+impl<'a, 'alloc, F, Hal> BivariateMLEcheckProver<'a, 'alloc, F, Hal>
+where
+	F: TowerField,
+	Hal: ComputeLayer<F>,
+{
+	#[allow(clippy::too_many_arguments)]
+	pub fn new(
```

Comment:
Please document the constructor, especially the meaning of the eq_ind_partial_evals and first_round_eval_1s args.